### PR TITLE
Add support for fma intrinsic

### PR DESCRIPTION
**Pull request by kevindevos**
This adds support for the fma (fused multiply-add) llvm intrinsic.

I'm unsure if i did the fma.opt unit test correctly, as i added fast math flags to some of the tests. However i do not understand why other floating point tests such as fadd do not fail even when fast math flags are not specified. Shouldn't they fail when one of the inputs to fadd could be NaN?

Overloading fm_poison in instr.cpp was a little tricky, i'm not sure if there is a better solution without code duplication.

### CODE REVIEW

File: tests/unit/fp/fma-error.opt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Name: target more poisonous than source
+; ERROR: Target is more poisonous than source for float %t
+%t = fma float %a, float %b, float %c
+  =>
+%t = poison
+
+Name: value mismatch (2.0 * 4.0) + 1.0 => 8.0
```
Comment (nunoplopes) [inline tests/unit/fp/fma-error.opt]: due to limitations in the unit test runner, only 1 failing test per file is supported.

File: ir/instr.cpp [hunk 2]
```diff
@@ -700,52 +721,74 @@ void TernaryOp::print(ostream &os) const {
...
   auto &bv = s[*b];
   auto &cv = s[*c];
-  function<expr(const expr&, const expr&, const expr&)> fn;
+  function<StateValue(const expr&, const expr&, const expr&)> fn;
 
   switch (op) {
   case FShl:
-    fn = [](auto a, auto b, auto c) {
-      return expr::fshl(a, b, c);
+    fn = [&](auto a, auto b, auto c) -> StateValue {
```
Comment (nunoplopes) [inline ir/instr.cpp]: & doesn't seem needed here

File: ir/instr.cpp [hunk 3]
```diff
@@ -700,52 +721,74 @@ void TernaryOp::print(ostream &os) const {
...
+      return { expr::fshl(a, b, c), true };
     };
     break;
+
   case FShr:
-    fn = [](auto a, auto b, auto c) {
-      return expr::fshr(a, b, c);
+    fn = [&](auto a, auto b, auto c) -> StateValue {
```
Comment (nunoplopes) [inline ir/instr.cpp]: idem

File: ir/instr.cpp [hunk 4]
```diff
@@ -700,52 +721,74 @@ void TernaryOp::print(ostream &os) const {
...
+      return { expr::fshr(a, b, c), true };
+    };
+    break;
+
+  case FMA:
+    fn = [&](auto a, auto b, auto c) -> StateValue{
```
Comment (nunoplopes) [inline ir/instr.cpp]: no space after StateValue

File: ir/instr.cpp [hunk 5]
```diff
@@ -700,52 +721,74 @@ void TernaryOp::print(ostream &os) const {
...
+      return fm_poison(s, a, b, c, [](expr &a, expr &b, expr &c) { 
+                                   return expr::fma(a, b, c); }, fmath, false);
     };
     break;
   }
 
   if (getType().isVectorType()) {
-    vector<StateValue> vals;
```
Comment (nunoplopes) [inline ir/instr.cpp]: this change seems unrelated

File: ir/instr.cpp [hunk 6]
```diff
@@ -231,6 +232,14 @@ static StateValue fm_poison(State &s, expr a, expr b,
   return { move(val), move(non_poison) };
 }
 
+static StateValue fm_poison(State &s, expr a, expr b,
+                            function<expr(expr&,expr&)> fn,
+                            FastMathFlags fmath, bool only_input) {
+  return fm_poison(s, a, b, expr(),
```
Comment (nunoplopes) [inline ir/instr.cpp]: expr() doesn't work here. It will make all non_poison expressions above to be null.

File: ir/instr.cpp [hunk 7]
```diff
@@ -680,6 +689,18 @@ unique_ptr<Instr> UnaryOp::dup(const string &suffix) const {
...
 
+TernaryOp::TernaryOp(Type &type, std::string &&name, Value &a, Value &b, Value &c, Op op,
+            FastMathFlags fmath)
+    : Instr(type, std::move(name)), a(&a), b(&b), c(&c), op(op), fmath(fmath) {
+  switch (op) {
+    case FShr:
+    case FShl:
+      assert(fmath.isNone());
+      break;
+    default:
```
Comment (nunoplopes) [inline ir/instr.cpp]: no default pls. That's a hazard waiting to explode in the future

File: tests/unit/fp/fma-error.opt (hunk 1 above)
Comment (kevindevos) [reply tests/unit/fp/fma-error.opt]: Fixed, split into 2 files, fma-error-mismatch.opt and fma-poison-error.opt.

File: ir/instr.cpp (hunk 2 above)
Comment (kevindevos) [reply ir/instr.cpp]: Fixed

File: ir/instr.cpp (hunk 3 above)
Comment (kevindevos) [reply ir/instr.cpp]: Fixed

File: ir/instr.cpp (hunk 5 above)
Comment (kevindevos) [reply ir/instr.cpp]: Changed back to how it was before. This was only a line swap for it to be consistent with BinOp::toSMT().

File: ir/instr.cpp (hunk 7 above)
Comment (kevindevos) [reply ir/instr.cpp]: Fixed, changed to a case.

File: ir/instr.cpp (hunk 6 above)
Comment (kevindevos) [reply ir/instr.cpp]: Fixed, but kept this the same. I experimented passing expr(true) but that gets simplified and gives the same result as expr(). Instead the passed expr c is tested with isValid() before doing any operation with it, and if expr() is passed, nothing is added to the non_poison expr.

**Comment by kevindevos** [comment]
Apologies, had to force push since i did something wrong on my side with git and the commit included some code already in master of alive.

File: ir/instr.cpp (hunk 6 above)
Comment (nunoplopes) [reply ir/instr.cpp]: That doesn't work because c might be null.
You need an extra argument, say "bool is_ternary = true".

Comment (nunoplopes) [reply ir/instr.cpp]: Also, in this line you need move() for a & b.

File: ir/instr.cpp [hunk 8]
```diff
@@ -680,6 +691,18 @@ unique_ptr<Instr> UnaryOp::dup(const string &suffix) const {
   return make_unique<UnaryOp>(getType(), getName() + suffix, *val, op, fmath);
 }
 
+TernaryOp::TernaryOp(Type &type, std::string &&name, Value &a, Value &b, Value &c, Op op,
```
Comment (nunoplopes) [inline ir/instr.cpp]: this line is clearly over 80 chars.
Also, no need for std:: there.

File: ir/instr.cpp [hunk 9]
```diff
@@ -680,6 +691,18 @@ unique_ptr<Instr> UnaryOp::dup(const string &suffix) const {
...
+            FastMathFlags fmath)
+    : Instr(type, std::move(name)), a(&a), b(&b), c(&c), op(op), fmath(fmath) {
```
Comment (nunoplopes) [inline ir/instr.cpp]: nor std:: here

File: ir/instr.cpp (hunk 6 above)
Comment (kevindevos) [reply ir/instr.cpp]: Thanks!, all fixed, i need to pay more attention.
